rsfbclient-core/src/charset.rs

```diff
@@ -0,0 +1,194 @@
+//! Charset definitions and functions
+//!
+//! [Reference](http://www.destructor.de/firebird/charsets.htm)
+
+use encoding::{all, types::EncodingRef, DecoderTrap, EncoderTrap};
+use std::str;
+
+use crate::FbError;
+
+/// Charset definition. Used to encode/decode the
+/// strings.
+pub struct Charset {
+    pub fb: &'static str,
+    pub str: Option<EncodingRef>,
+}
+
+impl Charset {
+    /// Decode the bytes using the current charset
+    pub fn decode(&self, bytes: &[u8]) -> Result<String, FbError> {
+        if let Some(charset) = self.str {
+            charset.decode(bytes, DecoderTrap::Strict).map_err(|e| {
+                format!(
+                    "Found column with an invalid {} string: {}",
+                    charset.name(),
+                    e
+                )
+                .into()
+            })
+        } else {
+            str::from_utf8(bytes)
+                .map(|str| str.to_string())
+                .map_err(|e| format!("Found column with an invalid UTF-8 string: {}", e).into())
+        }
+    }
+
+    // Encode the string into bytes using the current charset
+    pub fn encode<S: Into<String>>(&self, str: S) -> Result<Vec<u8>, FbError> {
```

Comment:
The trait `Into<T>` here and the corresponding `&str.into()` are very handy for use, but theyt incur in a small penalty because of the dynamic dispatch that they use.

Maybe we could have a overloaded/alternative function using `&str` that could be used at leas in internal code for high performance?

``` rust
pub fn encode_str(&self, str: &str) -> Result<Vec<u8>, FbError> {
    // ...
}
```

Reply:
While i do agree that the `Into<String>` here is not necessary, rust generics do not cause dynamic dispatch, they work by creating multiple versions of the function for the necessary types, so it has only compiletime overhead.
Dynamic dispatch would be used if the type of the str variable was `Box<dyn Into<String>>` for example.

Reply:
Ok, rust monomorphization will handle all this during the compiler phase.